File: src/day2.js
'use strict';

const fs = require('fs');

function part1() {
	const lines = fs.readFileSync('./input.txt').toString().split('\r\n');

	let num = 0;

	for(const line of lines) {
		const parts = line.split(' ');

		const digits = parts[0].split('-');
		const min = Number(digits[0]);
		const max = Number(digits[1]);

		const letter = parts[1].charAt(0);
		let counter = 0;

		for(let i = 0; i < parts[2].length; i++) {
			if(parts[2].charAt(i) === letter) {
				counter++;
			}
		}

		if(counter >= min && counter <= max) {
			num++;
		}
	}
	return num;
}

function part2() {
	const lines = fs.readFileSync('./input.txt').toString().split('\r\n');

	let num = 0;

	for(const line of lines) {
		const parts = line.split(' ');

		const digits = parts[0].split('-');

		const first = Number(digits[0]);
		const second = Number(digits[1]);

		const letter = parts[1].charAt(0);
		let counter = 0;

		if(parts[2].charAt(first - 1) === letter) {
			counter++;
		}

		if(parts[2].charAt(second - 1) === letter) {
			counter++;
		}

		if(counter === 1) {
			num++;
		}
	}
	return num;
}

console.log(part1());
console.log(part2());

File: 2017/day2.clj
(ns clojure-aoc.core (:gen-class))

(defn diff
  [line]
  (do
    (def parts (clojure.string/split line #"\t"))
    (def nums (into [] (map #(Integer/parseInt %) parts)))
    (- (apply max nums) (apply min nums))
  )
)

(defn div
  [line]
  (do
    (def parts (clojure.string/split line #"\t"))
    (def nums (into [] (map #(Integer/parseInt %) parts)))
    (loop [i 0] (when (< i (count nums))
      (let [ret (loop [j (+ i 1)] (when (< j (count nums))
        (let [
          [x y] (map (partial get nums) [i j])
          big (max x y)
          small (min x y)
        ]
          (if (= (mod big small) 0)
            (/ big small)
            (recur (inc j))
          )
        )
      ))]
        (if (nil? ret)
          (recur (inc i))
          ret
        )
      )
    ))
  )
)

(defn -main
  [& args]
  (do
    (def input (slurp "input.txt"))
    (def part1 (reduce + (map #(diff %) (clojure.string/split input #"\n"))))
    (def part2 (reduce + (map #(div %) (clojure.string/split input #"\n"))))
    (println (str part1 " " part2))
  )
)
